### PR TITLE
feature: US729482 Create COBOL LS VSC extension public API #801

### PR DESCRIPTION
**Pull request by Nurkambay**
1. Implementation: Populate public extended API let other plugins use COBOL LS plugin features
1. Unit tests

### CODE REVIEW

**Comment by temanbrcom** [comment]
Please, remove the user story number from the commit message

File: clients/cobol-lsp-vscode-extension/src/extension.ts [hunk 1]
```diff
@@ -32,6 +32,8 @@ import {TelemetryService} from "./services/reporter/TelemetryService";
 import {createFileWithGivenPath} from "./services/Settings";
 import {ZoweApi} from "./services/ZoweApi";
 import {resolveSubroutineURI} from "./services/util/SubroutineUtils";
+import { AnalysisResult } from "./api/model/AnalysisResult";
+import { Console } from "console";
```
Comment (grianbrcom) [inline clients/cobol-lsp-vscode-extension/src/extension.ts:36]: I guess that "console" is an unused import.

File: clients/cobol-lsp-vscode-extension/src/api/model/AnalysisResult.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import { Location } from "vscode";
+
+export class AnalysisResult {
+    public paragraphRange: Map<string, Location[]>;
+    public paragraphUsages: Map<string, Location[]>;
+    public sectionRange: Map<string, Location[]>;
+    public sectionUsages: Map<string, Location[]>;
```
Comment (grianbrcom) [inline clients/cobol-lsp-vscode-extension/src/api/model/AnalysisResult.ts:21]: The result is something in-between state. It's not a full AnalysisResult received from the Java side and not a structure that CCF expects from our plugin. I'll reach you personally to discuss.

**Review by grianbrcom — APPROVED** [review]
Looks good, but not that I expect)

File: clients/cobol-lsp-vscode-extension/src/api/ExtendedApi.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import { LanguageClientService } from "../services/LanguageClientService";
```
Comment (ishche) [inline clients/cobol-lsp-vscode-extension/src/api/ExtendedApi.ts:15]: There is no need for a class. I think it's better to remove it.

File: clients/cobol-lsp-vscode-extension/src/api/model/AnalysisResult.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import { Location } from "vscode";
+
+export class AnalysisResult {
```
Comment (ishche) [inline clients/cobol-lsp-vscode-extension/src/api/model/AnalysisResult.ts:17]: This class can be replaced with a type.

File: clients/cobol-lsp-vscode-extension/src/api/model/RetrieveRangeEvent.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+export class RetrieveRangeEvent {
+    public uri: string;
```
Comment (ishche) [inline clients/cobol-lsp-vscode-extension/src/api/model/RetrieveRangeEvent.ts:16]: Same here, no need for a class

File: clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts [hunk 6]
```diff
@@ -53,6 +55,14 @@ export class LanguageClientService {
         this.handlers.push(languageClient => languageClient.onRequest(method, handler));
     }
 
+    public retrieveAnalysis(uri: string): Promise<AnalysisResult> {
+        const languageClient = this.getLanguageClient();
+        const request: RetrieveRangeEvent = new RetrieveRangeEvent();
+        request.uri = uri;
+
+        return Promise.resolve(languageClient.sendRequest("extended/analysis", request));
```
Comment (ishche) [inline clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts:63]: Please consider async/await syntax instead

File: clients/cobol-lsp-vscode-extension/test-report.xml [hunk 7]
```diff
@@ -0,0 +1,187 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ishche) [inline clients/cobol-lsp-vscode-extension/test-report.xml:1]: Probably this file is here by mistake.

**Review by ishche — CHANGES_REQUESTED** [review]

File: clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    vscode.workspace.getConfiguration(expect.any(String)).get = jest.fn().mockReturnValue(0);
+});
+
+describe("Check ExtendedApi class functionality", () => {
+
+    beforeEach(() => {
+        languageClientService = new LanguageClientService(middleware);
+    });
+
+    test("ExtendedApi rerurns AnalysisResult", async() => {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts:58]: rerurns ?

File: clients/cobol-lsp-vscode-extension/test-report.xml (hunk 7 above)
Comment (asatklichov) [reply clients/cobol-lsp-vscode-extension/test-report.xml:1]: Yes, this file must not be included in repo

File: clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+const copyBooksDownloader: CopybookDownloadService = new CopybookDownloadService(zoweApi, null, null);
+const middleware: Middleware = new Middleware(new CopybookURI(profileService), copyBooksDownloader);
+let languageClientService: LanguageClientService;
+
+beforeEach(() => {
+    jest.clearAllMocks();
+    vscode.workspace.getConfiguration(expect.any(String)).get = jest.fn().mockReturnValue(0);
+});
+
+describe("Check ExtendedApi class functionality", () => {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts:52]: You have one describe suite, but two beforeEach(), why?

File: clients/cobol-lsp-vscode-extension/test-report.xml (hunk 7 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/test-report.xml]: Fixed!

File: clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts (hunk 8 above)
Comment (Nurkambay) [reply clients/cobol-lsp-vscode-extension/src/__tests__/ExtendedApiTest.ts]: Fixed

**Review by temanbrcom — CHANGES_REQUESTED** [review]
Looks ok, just remove the "US729482" from the commit message - it is an internal number and it makes no sense to expose it

File: clients/cobol-lsp-vscode-extension/src/__tests__/extensionTest.spec.ts [hunk 10]
```diff
@@ -93,6 +93,12 @@ describe("Check plugin extension for cobol starts successfully.", () => {
         .toBeCalledWith({scheme: "file", language: "COBOL"}, expect.any(CopybooksCodeActionProvider));
 
     });
+
+    test("extension rerurns extended API surface", async() => {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/__tests__/extensionTest.spec.ts:97]: type here: should be reruns

**Review by temanbrcom — APPROVED** [review]
🚗

**Review by asatklichov — APPROVED** [review]
🖌️

**Pull request merged**